dlobron/ChugBot
# Virtual Attendance System

### Description
Creates a way for chug leaders to use ChugBot to see live, up to date attendance sheets and take attendance for the chugim they are running. Also provides a mechanism for roshes/yoetzot to review attendance and quickly communicate which campers are missing to their staff, as well as track some attendance trends throughout the summer. The administrator must designate which time blocks are active for a given chug group, and then other tzevet may begin taking attendance.

### Specific Features
- Creates chug leader and rosh edah logins, enabling access to virtual attendance system
    - Customizes the navigation bar for different user-access levels
- Chug leader searches for edah/edot and chug through a simple form on their own designated home page
- Chug leader can press checkmarks to then indicate which campers are present (with a toggle all button available), and save their selections using a button at the base of the screen
- Rosh/yoetzet has option to view the attendance of one perek (default) or detailed matrix with historical data
    - Regardless, searches for the specific options through a similarly simple/intuitive form
- Perek Attendance View: Rosh/yoetzet sees list of campers sorted by edah, bunk, and then name. Also shows what chug they are supposed to be in and their current status (present, absent, attendance not yet taken) both by color and an icon
    - Ability to press a button below the attendance to automatically copy a simply-formatted plaintext attendance report to device clipboard, facilitating easy communication to other members of the staff
- Matrix Attendance View: Rosh/yoetzet sees list of campers in one edah with their historic attendance records (bound by provided dates) to identify trends in missing campers
    - One row per camper with one column per date, and an icon in the corresponding cell indicating the attendance of that camper on that date
    - 1 attendance matrix per chug group with attendance records within the provided dates
- Admin ability to designate which block is active for a chug group (and therefore ability to release chug lists once they are finalized)
- Admin ability to purge old attendance records (e.g. delete ones from previous sessions or summers)
- Other minor changes/updates as needed

### Testing Instructions
In the directory where the repo is stored locally, run this command:
```
./start.sh
```
Then, navigate to [http://127.0.0.1:8000/](http://127.0.0.1:8000/) on that device and test normally. No coding should be required for basic testing.

By default, the testing database sets the rosh password to "developerrosh" and the chug leader password to "developerchug" (though can [and for thorough testing, should] be changed through the main admin settings page) All chug groups are set to have "Session A" as the active block. No attendance records are provided; the tester must "take attendance" from scratch.

### Installation/Upgrade Instructions
This update requires 2 new tables in the database:
- `attendance` - creates a record for every camper who has attendance taken (`camper_id`, `date`, `chug_instance_id`, `attendance_id`, `present`)
- `attendance_block_by_date` - identifies the block which is active for a particular date and chug group, keeping a record of information to accurately retrieve historical attendance lists
It also requires a few additional columns in other tables:
- `active_block_id` in `chug_groups` table tracks the block which is active for a particular chug group (perek)
- `chug_leader_password` and `rosh_yoetzet_password` in `admin_data` hold hashed passwords to be matched against their respective roles.

To complete these updates on the live databases of the real, running application, sign into the SQL server, select the correct database, and run the following to create the structure which will be populated as the application runs:
```
--
-- Table structure for table `attendance`
--

DROP TABLE IF EXISTS `attendance`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `attendance` (
  `camper_id` int NOT NULL,
  `date` date NOT NULL,
  `chug_instance_id` int NOT NULL,
  `attendance_id` int NOT NULL AUTO_INCREMENT,
  `present` tinyint NOT NULL,
  PRIMARY KEY (`attendance_id`),
  UNIQUE KEY `uk_attendance_present` (`camper_id`,`date`,`chug_instance_id`),
  CONSTRAINT `attendance_present_ibfk_1` FOREIGN KEY (`camper_id`) REFERENCES `campers` (`camper_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `attendance_present_ibfk_2` FOREIGN KEY (`chug_instance_id`) REFERENCES `chug_instances` (`chug_instance_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;


--
-- Table structure for table `attendance_block_by_date`
--

DROP TABLE IF EXISTS `attendance_block_by_date`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `attendance_block_by_date` (
  `date` date NOT NULL,
  `group_id` int NOT NULL,
  `block_id` int NOT NULL,
  `attendance_block_by_date_id` int NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`attendance_block_by_date_id`),
  UNIQUE KEY `uk_attendance_block_by_date` (`date`,`group_id`),
  CONSTRAINT `attendance_block_by_date_ibfk_1` FOREIGN KEY (`group_id`) REFERENCES `chug_groups` (`group_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `attendance_block_by_date_ibfk_2` FOREIGN KEY (`block_id`) REFERENCES `blocks` (`block_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;

alter table chug_groups add active_block_id INTEGER, add constraint foreign key(active_block_id) references blocks(block_id);

alter table admin_data add chug_leader_password varchar(255), add rosh_yoetzet_password varchar(255);
```
